### PR TITLE
Lazy-loading images

### LINKED ISSUE #673: Implement lazy-loading images
**Is your feature request related to a problem? Please describe.**
Following the POC in https://github.com/bbc/simorgh/issues/520 we should implement the `react-lazyload` library for our images.

**Describe the solution you'd like**
Implement the `<LazyLoad>` component (via `react-lazyload`) without a placeholder and with an offset (suggesting 250 but this could be variable), also ensure the prop `once` is passed. 

EG:
in `/src/app/containers/Figure/index.jsx`
```
      <LazyLoad offset={250} once>
        <Image alt={alt} src={src} />
      </LazyLoad>
```

Note: the offset value should be determined by pairing with UX designer - since the offset value is relating to when the lazyloading will start - e.g. `250` is two and a half viewport heights. 

Unit tests should be added with enzyme - we shouldn't just use snapshot unit testing. 
We should investigate whether Cypress tests can be used to test the lazy-loading is working correctly. 

When working on this ticket, we should pair with a UX Designer. 

**Describe alternatives you've considered**
N/A

*Testing Notes*
This will require performance testing to make sure that the images load as expected over slower than usual connections (Edge, 2G, etc.).

### PR DESCRIPTION
**Pull request by 12**
Resolves #673 

**Overall change:** Implements lazy-loading capabilities for the `Figure` container.

**Code changes:**

- Adds `react-lazyload` dependency
- Lazy-load images in the `Figure` container

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/Figure/index.jsx [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+import { string, number, objectOf, any, bool } from 'prop-types';
 import Figure from '@bbc/psammead-figure';
 import Image, { AmpImg } from '@bbc/psammead-image';
 import ImagePlaceholder from '@bbc/psammead-image-placeholder';
+import LazyLoad from 'react-lazyload';
 import Copyright from '../Copyright';
 import Caption from '../Caption';
 import { RequestContext } from '../../contexts/RequestContext';
 
+const LAZYLOAD_OFFSET = 250;
```
Comment (pjlee11) [inline src/app/containers/Figure/index.jsx]:
```suggestion
const LAZYLOAD_OFFSET = 250; // amount of pixels below the viewport to begin loading the image
```

File: src/app/containers/Figure/index.jsx [hunk 2]
```diff
@@ -50,17 +69,19 @@ FigureContainer.propTypes = {
   alt: string.isRequired,
   captionBlock: objectOf(any),
   copyright: string,
+  height: number,
+  lazyload: bool,
```
Comment (pjlee11) [inline src/app/containers/Figure/index.jsx]: Not actually asking for changes but should `lazyload` actually be `lazyLoad`

File: cypress/integration/newsBodySpec.js [hunk 3]
```diff
@@ -48,7 +48,25 @@ describe('Article Body Tests', () => {
   });
 
   it('should have a visible image with a caption', () => {
```
Comment (pjlee11) [inline cypress/integration/newsBodySpec.js]:
```suggestion
  it('should have a visible image with a caption that is lazyloaded and has a noscript fallback image', () => {
```

File: src/app/containers/Figure/fixtureData.jsx [hunk 4]
```diff
@@ -150,34 +156,39 @@ const generateFixtureData = ({ caption, copyright, platform, type }) => (
...
   caption: null,
   copyright: null,
+  lazyLoad: false,
   platform: 'canonical',
   type: '',
 };
 
 export const FigureImage = generateFixtureData({ platform: 'canonical' });
 
+export const FigureLazyLoadImage = generateFixtureData({ lazyLoad: true });
```
Comment (sareh) [inline src/app/containers/Figure/fixtureData.jsx]: You also need to pass in `platform` - since it's a required prop of `RequestContextProvider`

File: src/app/containers/Figure/index.jsx [hunk 5]
```diff
@@ -1,12 +1,28 @@
...
 
+const LAZYLOAD_OFFSET = 250; // amount of pixels below the viewport to begin loading the image
+
+const renderImage = (imageToRender, lazyLoad) =>
+  lazyLoad ? (
+    <Fragment>
+      <LazyLoad offset={LAZYLOAD_OFFSET} once>
+        {imageToRender}
+      </LazyLoad>
+
```
Comment (sareh) [inline src/app/containers/Figure/index.jsx]:
```suggestion
```

File: src/app/containers/Image/index.jsx [hunk 6]
```diff
@@ -77,6 +77,7 @@ const ImageContainer = ({ blocks }) => {
         ratio={ratio}
         src={rawImageSrc}
         width={width}
+        lazyLoad
```
Comment (sareh) [inline src/app/containers/Image/index.jsx:80]: We should check with UX whether they want every image lazyloaded.

**Comment by 12** [comment]
The offset of `250px` has been confirmed by @nataliesmart. We've agreed to create an issue to investigate an animation to ease the image into the page once it has fully loaded.

File: src/app/containers/Image/index.jsx (hunk 6 above)
Comment (sareh) [reply src/app/containers/Image/index.jsx:80]: I note that you've opened up this issue: https://github.com/bbc/simorgh/issues/1765 
Are UX & Product happy about this work happening in a separate PR - so that in the interim between this PR being merged and the ticket being completed, all images on live articles will be lazyloaded?

**Comment by sareh** [comment]
The implementation looks great. Great work, @rhell.
I'd just want to check with @jimjohnsonrollings if he'll be okay with this lazyloading experience for the first image existing for a while until we carry out this issue #1765
lazyloaded image at top of page

**Review by sareh — APPROVED** [review]

**Comment by 12** [comment]
Thanks @sareh, will be confirming with @jimjohnsonrollings and @nataliesmart today. Will post feedback.

**Comment by 12** [comment]
```
Jim J-R [12:13]
My thoughts on the lazy loading:
- Current PR is fine to go out
- I’d like some sort of metric that can say how much better the page is after doing this work
- We should look at not-lazy loading any image within the first three blocks of the article (as a proxy for ‘above the fold’)
- We should add the fade as an enhancement
```

**Comment by 12** [comment]
**With Lazy Loading**
image

**Without Lazy Loading**
image

**Review by pjlee11 — COMMENTED** [review]
I approve this PR, but because I was involved I won't give it an official review so we get 2 sets of fresh eyes for official reviews. Great work 👍

File: cypress/integration/newsBodySpec.js [hunk 7]
```diff
@@ -47,8 +47,26 @@ describe('Article Body Tests', () => {
     visibleImageNoCaption(getElement('figure').eq(0));
   });
 
-  it('should have a visible image with a caption', () => {
-    visibleImageWithCaption(getElement('figure').eq(2));
+  it('should have a visible image with a caption that is lazyloaded and has a noscript fallback image', () => {
+    let thirdFigure;
```
Comment (oluoluoxenfree) [inline cypress/integration/newsBodySpec.js:51]: Why third?

Comment (12) [reply cypress/integration/newsBodySpec.js:51]: Because we can guarantee this is below the fold, which means that we can 
1 - more reliably test that the image has not loaded when the page loads initially, and instead has an empty placeholder
2 - more reliably test that when scrolling the page, the empty placeholder is replaced with the actual image

**Review by oluoluoxenfree — APPROVED** [review]

**Comment by oluoluoxenfree** [comment]
Great work 💯

**Comment by jamesbrumpton** [comment]
LGTM. Happy for this to be merged.
